packet/bgp/bgp.go

```diff
@@ -5103,25 +5098,31 @@ func (p *PathAttributeMpReachNLRI) Serialize() ([]byte, error) {
 	nexthoplen := 4
 	if afi == AFI_IP6 {
 		nexthoplen = 16
-		if p.LinkLocalNexthop != nil {
-			nexthoplen += 16
-		}
 	}
 	offset := 0
 	switch safi {
 	case SAFI_MPLS_VPN:
 		offset = 8
-		nexthoplen += 8
+		nexthoplen += offset
 	case SAFI_FLOW_SPEC_VPN, SAFI_FLOW_SPEC_UNICAST:
 		nexthoplen = 0
 	}
+	if p.LinkLocalNexthop != nil {
+		nexthoplen *= 2
+	}
 	buf := make([]byte, 4+nexthoplen)
 	binary.BigEndian.PutUint16(buf[0:], afi)
 	buf[2] = safi
 	buf[3] = uint8(nexthoplen)
-	copy(buf[4+offset:], p.Nexthop)
-	if p.LinkLocalNexthop != nil {
-		copy(buf[4+offset+len(p.Nexthop):], p.LinkLocalNexthop)
+	if nexthoplen != 0 {
+		if afi == AFI_IP6 {
+			copy(buf[4+offset:], p.Nexthop.To16())
```

Comment:
Thank you for reviewing my patch.

Exactly.
I didn't mention enough.

To fix only conversion of IPv4 to IPv6, using `To16()` is sufficient.
But when testing it, I found that `nexthoplen` calculation is also need to parse and serialize NextHop field, otherwise "mpreach nexthop length is incorrect" at GoBGP or "Invalid NextHop length" at  BGP peers.

Should I separate this patch at this point? Or need to mention more clearly?